test/restart/mzcompose.py

```diff
@@ -332,6 +336,76 @@ def workflow_allowed_cluster_replica_sizes(c: Composition) -> None:
     )
 
 
+def workflow_allow_user_sessions(c: Composition) -> None:
+    c.up("materialized")
+    http_port = c.port("materialized", 6876)
+
+    # Ensure new user sessions are allowed.
+    c.sql(
+        "ALTER SYSTEM SET allow_user_sessions = true",
+        port=6877,
+        user="mz_system",
+    )
+
+    # SQL and HTTP user sessions should work.
+    assert c.sql_query("SELECT 1") == ([1],)
+    assert requests.post(
+        f"http://localhost:{http_port}/api/sql", json={"query": "select 1"}
+    ).json()["results"][0]["rows"] == [["1"]]
+
+    # Save a cursor for later.
+    cursor = c.sql_cursor()
+
+    # Disallow new user sessions.
+    c.sql(
+        "ALTER SYSTEM SET allow_user_sessions = false",
+        port=6877,
+        user="mz_system",
+    )
+
+    # New SQL and HTTP user sessions should now fail.
```

Comment:
Maybe add a test that a system user can still log in?

Reply:
That happens on L392 when we log in as `mz_system` to twiddle `allow_user_sessions` back to true!